Automattic/wp-calypso
# Checkout: Extract web pay mechanisms from Apple Pay

#### Changes proposed in this Pull Request

This PR refactors the Apple Pay payment method and payment processor to separate out the code that's generic between Apple Pay and Google Pay (together called "Web Pay"). It will make it considerably easier to add support for Google Pay in https://github.com/Automattic/wp-calypso/pull/52311. This PR should have no effect on the system.

The specific changes made by this PR are:

- Rename `useIsApplePayAvailable` to `useIsWebPayAvailable` and refactor it to be more generic.
- Rename the `applePayProcessor` to `webPayProcessor`.
- Move `usePaymentRequestOptions` and `useStripePaymentRequest` from the Apple Pay payment method to a utility file.
- Refactor `usePaymentRequestOptions` and `useStripePaymentRequest` to be more generic.
- Move the `APPLE_PAY_TRANSACTION_BEGIN` analytics event from the Apple Pay payment method into `webPayProcessor`.
- Translate the payment method slug `WPCOM_Billing_Web_Payment` into `web-pay` instead of `apple-pay` and vice versa.
- Move some filtering from `filterAppropriatePaymentMethods` to `isPaymentMethodEnabled`.
- Fix a bug which incorrectly failed to block Ebanx TEF when the country code is not BR.
- Stop passing unnecessary data (total, line items) to `webPayProcessor`.
- Throw an error if the Apple Pay payment method is not available when the button loads instead of logging an error and showing a disabled button. This state should never occur if `useIsWebPayAvailable` is working, so throwing an error will bring it to our attention faster if it does.
- Remove an old guard for a bug in old versions of Safari that has been fixed for 2 years and which I think is not relevant since we switched to using Stripe for Apple Pay.

#### Testing instructions

- Make a purchase with Apple Pay and verify that it works as expected. Instructions for doing this are complex; please contact me if you have questions.
- Verify that the `calypso_checkout_composite_apple_pay_submit_clicked` Tracks event is fired during checkout.

We should also verify that the Ebanx TEF payment method is still available because this touches that filtering code.

- Apply D47729-code to force the server to make the payment method returned by the server.
- Visit checkout with a product in the cart.
- Select Brazil as a country in the contact form.
- Verify that "Transferência bancária" is available as a payment method.
- Select another country in the contact form.
- Verify that "Transferencia bancaria" is NOT available as a payment method.
